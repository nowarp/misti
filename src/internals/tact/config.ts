import { ExecutionException, throwZodError } from "../exceptions";
import { ProjectName } from "../ir";
import {
  Config as TactConfig,
  ConfigProject,
  parseConfig,
} from "@tact-lang/compiler/dist/config/parseConfig";
import fs from "fs";
import path from "path";

/**
 * Manages the logic around the Tact configuration file.
 *
 * Tact config describes the structure of the project, and includes the entry
 * points to run compilation and analysis on.
 */
export class TactConfigManager {
  private constructor(
    /**
     * An absolute path to the root directory storing the configuration file.
     *
     * If the config is generated for the Tact contract, it should be a directory containing all the imported files.
     */
    private projectRoot: string,
    /** Tact config parsed with Zod. */
    private config: TactConfig,
  ) {}

  /**
   * Creates a TactConfigManager from a Tact configuration file typically specified by the user.
   *
   * @param ctx Misti context.
   * @param tactConfigPath Path to the Tact configuration file.
   */
  public static fromConfig(tactConfigPath: string): TactConfigManager {
    return new TactConfigManager(
      path.resolve(path.dirname(tactConfigPath)),
      this.readTactConfig(tactConfigPath),
    );
  }

  /**
   * Creates a TactConfigManager from a single Tact contract.
   *
   * @param ctx Misti context.
   * @param projectName Name of the project.
   * @param contractPath Path to the Tact contract.
   */
  public static fromContract(
    projectRoot: string,
    contractPath: string,
    projectName: ProjectName = path.basename(
      contractPath,
      ".tact",
    ) as ProjectName,
  ): TactConfigManager {
    const tactConfig: TactConfig = {
      projects: [
        {
          name: projectName,
          path: path.relative(projectRoot, contractPath),
          output: "/tmp/misti/output", // never used
          options: {
            debug: false,
            external: true,
          },
        },
      ],
    };
    return new TactConfigManager(path.resolve(projectRoot), tactConfig);
  }

  public getConfig(): TactConfig {
    return this.config;
  }

  /**
   * Returns absolute path to the project root.
   */
  public getProjectRoot(): string {
    return this.projectRoot;
  }

  /**
   * Gets projects defined within the configuration file.
   */
  public getProjects(): ConfigProject[] {
    return this.config.projects;
  }

  /**
   * Find the project config based on the provided name.
   */
  public findProjectByName(
    projectName: ProjectName,
  ): ConfigProject | undefined {
    return this.config.projects.find((project) => projectName === project.name);
  }

  /**
   * Find the project config based on the provided path.
   */
  public findProjectByPath(projectPath: string): ConfigProject | undefined {
    return this.config.projects.find(
      (project) => projectPath === this.resolveProjectPath(project.path),
    );
  }

  /**
   * Returns an absolute path or the project based on the project path.
   */
  public resolveProjectPath(projectPath: string): string {
    return path.resolve(this.projectRoot, projectPath);
  }

  /**
   * Reads the Tact configuration file from the specified path, parses it, and returns
   * the TactConfig object.
   * @throws If the config file does not exist or cannot be parsed.
   * @returns The parsed TactConfig object.
   */
  private static readTactConfig(tactConfigPath: string): TactConfig {
    const resolvedPath = path.resolve(tactConfigPath);
    if (!fs.existsSync(resolvedPath)) {
      throw ExecutionException.make(
        `Unable to find config file at ${resolvedPath}`,
      );
    }
    try {
      return parseConfig(fs.readFileSync(resolvedPath, "utf8"));
    } catch (err) {
      throwZodError(err, {
        msg: `Incorrect Tact Project file ${resolvedPath}:`,
        help: [
          `Ensure ${resolvedPath} is a Tact Project file.`,
          "See https://docs.tact-lang.org/book/config/ for additional information.",
        ].join(" "),
      });
    }
  }

  /**
   * Returns absolute paths to entry points specified in the Tact configuration file.
   */
  public getEntryPoints(): string[] {
    return this.config.projects.map((p) => this.resolveProjectPath(p.path));
  }
}
