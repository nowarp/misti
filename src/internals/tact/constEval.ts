import { evalConstantExpression } from "@tact-lang/compiler/dist/constEval";
import { CompilerContext } from "@tact-lang/compiler/dist/context";
import { AstExpression } from "@tact-lang/compiler/dist/grammar/ast";
import {
  Value,
  StructValue,
  CommentValue,
} from "@tact-lang/compiler/dist/types/types";
import { Address, Cell, Slice } from "@ton/core";

/**
 * Evaluates a constant expression and returns its value.
 *
 * @param expr The AST expression to evaluate.
 * @returns The evaluated constant value, or undefined if evaluation fails.
 */
export function evalExpr(expr: AstExpression): Value | undefined {
  try {
    return evalConstantExpression(expr, new CompilerContext());
  } catch {
    return undefined;
  }
}

/**
 * Evaluates the given expression to a constant value and checks if it matches
 * the expected type.
 *
 * @param expr The expression to evaluate.
 * @param expectedType The expected type name as a string.
 * @returns The evaluated value if it matches the expected type, undefined otherwise.
 */
export function evalToType(
  expr: AstExpression,
  expectedType: string,
): Value | undefined {
  const value = evalExpr(expr);
  return value !== undefined && checkType(value, expectedType)
    ? value
    : undefined;
}

/**
 * Evaluates the given expression to a constant value and checks if it matches
 * the expected type and value.
 *
 * @param expr The expression to evaluate.
 * @param expectedType The expected type name as a string.
 * @param expectedValue The expected value.
 * @returns True if the expression can be evaluated to a constant value that
 *          matches the expected type and value, false otherwise.
 */
export function evalsToValue(
  expr: AstExpression,
  expectedType: string,
  expectedValue: Value,
): boolean {
  const value = evalToType(expr, expectedType);
  return value !== undefined && value === expectedValue;
}

/**
 * Evaluates the given expression to a constant value and checks if it satisfies the predicate.
 *
 * @param expr The expression to evaluate.
 * @param predicate The predicate to check.
 * @returns True if the expression can be evaluated to a constant value that satisfies
 *          the predicate, false otherwise.
 */
export function evalsToPredicate(
  expr: AstExpression,
  predicate: (value: any) => boolean,
): boolean {
  const value = evalExpr(expr);
  return value !== undefined && predicate(value);
}

function checkType(value: Value, expectedType: string): boolean {
  switch (expectedType) {
    case "bigint":
      return typeof value === "bigint";
    case "boolean":
      return typeof value === "boolean";
    case "string":
      return typeof value === "string";
    case "null":
      return value === null;
    case "Address":
      return value instanceof Address;
    case "Cell":
      return value instanceof Cell;
    case "Slice":
      return value instanceof Slice;
    case "CommentValue":
      return value instanceof CommentValue;
    case "StructValue":
      return isStructValue(value);
    default:
      return false;
  }
}

function isStructValue(value: any): value is StructValue {
  return (
    typeof value === "object" &&
    value !== null &&
    !(value instanceof CommentValue) &&
    !Address.isAddress(value) &&
    !(value instanceof Cell) &&
    !(value instanceof Slice)
  );
}
