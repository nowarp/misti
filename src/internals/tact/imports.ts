export {
  AstStatementCondition,
  AstAsmFunctionDef,
  AstOpBinary,
  AstConstantDef,
  AstConstantDecl,
  AstContract,
  AstContractDeclaration,
  AstContractInit,
  AstExpression,
  AstFieldAccess,
  AstFieldDecl,
  AstFunctionDef,
  AstFunctionDecl,
  AstId,
  AstImport,
  AstMessageDecl,
  AstMethodCall,
  AstModule,
  AstNativeFunctionDecl,
  AstNode,
  AstNumberBase,
  AstPrimitiveTypeDecl,
  AstReceiver,
  AstStatement,
  AstStaticCall,
  AstString,
  AstStructDecl,
  AstStructFieldInitializer,
  AstStructInstance,
  AstTrait,
  AstTraitDeclaration,
  AstType,
  AstStatementLet,
  AstStatementAssign,
  AstTypeDecl,
  AstNumber,
  AstStatementReturn,
  AstOpUnary,
  AstConditional,
  AstTypedParameter,
  AstOptionalType,
  AstStatementExpression,
  AstLiteral,
  AstFunctionAttributeName,
  prettyPrint,
  idText,
  isSelfId,
  tryExtractPath,
  isLiteral,
  getAstFactory,
  eqExpressions,
  getAstUtil,
  ItemOrigin,
  Source,
  asString as importAsString,
  SrcInfo,
  dummySrcInfo,
  getParser,
  srcInfoEqual,
  precompile,
  enableFeatures,
  stdLibFiles,
  AstStore,
  getRawAST,
  CompilerContext,
  Config,
  Project,
  parseConfig,
  evalConstantExpression,
  ensureInt,
} from "@tact-lang/compiler";
