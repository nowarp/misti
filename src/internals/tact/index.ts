export * from "./iterators";
export * from "./util";
export * from "./constEval";
export * from "./stdlib";
export * from "./config";
