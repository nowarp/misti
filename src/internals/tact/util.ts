import { forEachExpression } from "./iterators";
import { unreachable } from "../util";
import { AstComparator } from "@tact-lang/compiler/dist/";
import {
  AstExpression,
  AstId,
  AstContractInit,
  AstFieldDecl,
  AstFunctionDef,
  AstReceiver,
  AstContract,
  AstFieldAccess,
  AstStatement,
  AstMethodCall,
  SrcInfo,
  isSelfId,
  idText,
  tryExtractPath,
  AstCondition,
} from "@tact-lang/compiler/dist/grammar/ast";
import { prettyPrint } from "@tact-lang/compiler/dist/prettyPrinter";
import { Interval as RawInterval } from "ohm-js";
import * as path from "path";

/**
 * Creates a concise string representation of `SrcInfo`.
 */
export function formatPosition(ref?: SrcInfo): string {
  if (!ref || !ref.file) {
    return "";
  }
  const relativeFilePath = path.relative(process.cwd(), ref.file);
  const lc = ref.interval.getLineAndColumn();
  return `${relativeFilePath}: ${lc}\n`;
}

/**
 * Returns the accessor name without the leading `self.` part.
 *
 * For example:
 * - `self.a` -> AstId(`a`)
 * - `self.a()` -> AstMethodCall(`a`)
 * - `self.object.f1` -> AstFieldAccess(`object.f1`)
 * - `nonSelf.a` -> undefined
 */
export function removeSelf(
  expr: AstExpression,
): AstId | AstFieldAccess | undefined {
  if (expr.kind === "method_call") {
    return removeSelf(expr.self);
  }
  if (expr.kind === "field_access") {
    if (isSelf(expr.aggregate)) {
      return expr.field;
    } else {
      const newAggregate = removeSelf(expr.aggregate);
      if (newAggregate !== undefined) {
        return {
          ...expr,
          aggregate: newAggregate,
        };
      }
    }
  }
  return undefined;
}

/**
 * @returns True for self identifiers: `self`.
 */
export function isSelf(expr: AstExpression): boolean {
  return expr.kind === "id" && isSelfId(expr);
}

/**
 * @returns True for self access expressions: `self.a`, `self.a.b`.
 */
export function isSelfAccess(expr: AstExpression): boolean {
  const path = tryExtractPath(expr);
  return path !== null && path.length > 1 && isSelfId(path[0]);
}

/**
 * @returns True iff `call` is a stdlib method mutating its receiver.
 */
export function isStdlibMutationMethod(call: AstMethodCall): boolean {
  // https://docs.tact-lang.org/book/maps
  const mapMutationOperations = ["set", "del"];
  // See: stdlib/std/cells.tact
  const builderMutationOperations = [
    "storeInt",
    "storeUint",
    "storeBool",
    "storeBit",
    "storeCoins",
    "storeRef",
    "storeSlice",
    "storeBuilder",
    "storeAddress",
    "storeMaybeRef",
  ];
  const methodName = idText(call.method);
  return (
    // Filter out contract calls e.g.: `self.set(/*...*/)`
    !isSelf(call.self) &&
    // TODO: This should be rewritten when we have types in AST
    (mapMutationOperations.includes(methodName) ||
      builderMutationOperations.includes(methodName))
  );
}

export type MutatedElement = AstId | AstFieldAccess;

/**
 * Collects mutations local or state mutations within the statements.
 *
 * @param The statement to analyze
 * @param flatStmts If true, only traverse statements at the current level without
 *                  going into nested statements. It should be used when calling this function
 *                  inside one of the iterators.
 * @returns Mutated fields and local identifiers, including nested fields of mutated structure instances
 */
export function collectMutations(
  stmt: AstStatement,
  { flatStmts = false }: Partial<{ flatStmts: boolean }> = {},
):
  | { mutatedFields: MutatedElement[]; mutatedLocals: MutatedElement[] }
  | undefined {
  const mutatedFields: MutatedElement[] = [];
  const mutatedLocals: MutatedElement[] = [];

  const handleMethodCallsMutations = (): void => {
    forEachExpression(
      stmt,
      (expr: AstExpression) => {
        if (expr.kind === "method_call" && isStdlibMutationMethod(expr)) {
          if (isSelfAccess(expr.self)) {
            // Field mutation
            const mutated = removeSelf(expr);
            if (mutated) {
              mutatedFields.push(mutated);
            }
          } else {
            // Local mutation
            if (expr.self.kind === "field_access" || expr.self.kind === "id")
              mutatedLocals.push(expr.self);
          }
        }
      },
      { flatStmts },
    );
  };
  handleMethodCallsMutations();

  const handleAssignmentMutations = (): void => {
    if (
      stmt.kind === "statement_assign" ||
      stmt.kind === "statement_augmentedassign"
    ) {
      const field = removeSelf(stmt.path);
      if (field) {
        // Field mutations
        mutatedFields.push(field);
      } else {
        // Local mutations
        const local = stmt.path;
        if (local.kind === "field_access" || local.kind === "id") {
          mutatedLocals.push(local);
        }
      }
    }
  };
  handleAssignmentMutations();

  return !mutatedFields.length && !mutatedLocals.length
    ? undefined
    : { mutatedFields, mutatedLocals };
}

/**
 * Collects names of the mutated elements.
 *
 * For example:
 * - a -> a
 * - self.a -> a
 * - self.object.f1 -> object
 */
export function mutationNames(items: MutatedElement[]): string[] {
  return items.flatMap((item) => {
    if (item.kind === "id") {
      return [item.text];
    } else if (item.kind === "field_access") {
      const path = tryExtractPath(item);
      return path && path.length >= 2 ? [path[0].text] : [];
    } else {
      return [];
    }
  });
}

/**
 * Set containing information about the locations with some additional information.
 * We need this, since `SrcInfo` objects cannot be trivially compared.
 */
export class SrcInfoSet<T> {
  private items: [T, SrcInfo][];

  constructor(pairs?: [T, SrcInfo][]) {
    this.items = [];
    if (pairs) {
      pairs.forEach((pair) => this.add(pair));
    }
  }

  add(item: [T, SrcInfo]) {
    if (!this.has(item)) {
      this.items.push(item);
    }
  }

  has(item: [T, SrcInfo]): boolean {
    return this.items.some((existingItem) =>
      this.equals(existingItem[1], item[1]),
    );
  }

  delete(item: [T, SrcInfo]): boolean {
    const index = this.items.findIndex((existingItem) =>
      this.equals(existingItem[1], item[1]),
    );
    if (index !== -1) {
      this.items.splice(index, 1);
      return true;
    }
    return false;
  }

  extract(): [T, SrcInfo][] {
    return this.items.slice();
  }

  private equals(srcInfo1: SrcInfo, srcInfo2: SrcInfo): boolean {
    return (
      srcInfo1.file === srcInfo2.file &&
      srcInfo1.contents === srcInfo2.contents &&
      this.compareIntervals(srcInfo1.interval, srcInfo2.interval) &&
      srcInfo1.origin === srcInfo2.origin
    );
  }

  private compareIntervals(
    interval1: RawInterval,
    interval2: RawInterval,
  ): boolean {
    return (
      interval1.sourceString === interval2.sourceString &&
      interval1.startIdx === interval2.startIdx &&
      interval1.endIdx === interval2.endIdx &&
      interval1.contents === interval2.contents
    );
  }
}

/**
 * Returns true iff the input expression represents a primitive literal.
 */
export function isPrimitiveLiteral(expr: AstExpression): boolean {
  return ["null", "boolean", "number", "string"].includes(expr.kind);
}

/**
 * Checks if the AST of two nodes is equal using the Tact AST comparison API.
 */
export function nodesAreEqual(
  node1: AstExpression | AstStatement,
  node2: AstExpression | AstStatement,
): boolean {
  return AstComparator.make({ sort: true, canonicalize: false }).compare(
    node1,
    node2,
  );
}

/**
 * Checks if the AST of two lists of statements is equal using the Tact AST comparison API.
 */
export function statementsAreEqual(
  stmts1: AstStatement[],
  stmts2: AstStatement[],
): boolean {
  if (stmts1.length !== stmts2.length) return false;
  return stmts1.every((stmt, i) => {
    return nodesAreEqual(stmt, stmts2[i]);
  });
}

/**
 * Collects declarations of all the contract fields.
 */
export function collectFields(
  contract: AstContract,
  { initialized = false }: Partial<{ initialized: boolean }> = {},
): Map<string, AstFieldDecl> {
  return contract.declarations.reduce((acc, decl) => {
    if (
      decl.kind === "field_decl" &&
      (!initialized || decl.initializer !== null)
    ) {
      acc.set(decl.name.text, decl);
    }
    return acc;
  }, new Map<string, AstFieldDecl>());
}

/**
 * Returns the human-readable name of the function.
 */
export function funName(
  fun: AstReceiver | AstContractInit | AstFunctionDef,
): string {
  switch (fun.kind) {
    case "contract_init":
      return "init";
    case "receiver":
      return prettyPrint(fun).split("\n")[0].slice(0, -3);
    case "function_def":
      return idText(fun.name);
    default:
      unreachable(fun);
  }
}

/**
 * Collects all the conditions from the conditional, including `if` and `else if` statements.
 */
export function collectConditions(
  node: AstCondition,
  { nonEmpty = false }: Partial<{ nonEmpty: boolean }> = {},
): AstExpression[] {
  const conditions: AstExpression[] = nonEmpty
    ? node.trueStatements.length > 0
      ? [node.condition]
      : []
    : [node.condition];
  if (node.elseif) {
    conditions.push(...collectConditions(node.elseif));
  }
  return conditions;
}
