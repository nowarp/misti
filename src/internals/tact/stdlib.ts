import { MistiContext } from "../context";
import { hasSubdirs } from "../util";
import { SrcInfo } from "@tact-lang/compiler/dist/grammar/ast";
import path from "path";

/**
 * A mandatory part of the file path to stdlib if using the default path.
 */
export const DEFAULT_STDLIB_PATH_ELEMENTS = [
  "node_modules",
  "@tact-lang",
  "compiler",
  "stdlib",
];

/**
 * Returns an absolute path to Tact stdlib distributed within the tact compiler
 * package.
 */
export function getStdlibPath() {
  return path.join(
    path.dirname(require.resolve("@tact-lang/compiler/package.json")),
    "stdlib",
  );
}

/**
 * Checks if a given location or file path is defined in the Tact stdlib.
 * @param ctx MistiContext object
 * @param locOrPath SrcInfo object or string file path
 * @returns boolean indicating if the location is in the stdlib
 */
export function definedInStdlib(
  ctx: MistiContext,
  locOrPath: SrcInfo | string,
): boolean {
  const stdlibPath = ctx.config.tactStdlibPath;
  const pathElements =
    stdlibPath === undefined
      ? DEFAULT_STDLIB_PATH_ELEMENTS
      : stdlibPath.split("/").filter((part) => part !== "");
  const filePath = typeof locOrPath === "string" ? locOrPath : locOrPath.file;

  return (
    filePath !== null &&
    (filePath.startsWith("@stdlib") || hasSubdirs(filePath, pathElements))
  );
}
