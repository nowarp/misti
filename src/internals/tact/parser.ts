import { getStdlibPath } from "./stdlib";
import { MistiContext } from "../context";
import { TactException } from "../exceptions";
import { ConfigProject } from "@tact-lang/compiler/dist/config/parseConfig";
import { CompilerContext } from "@tact-lang/compiler/dist/context";
import { getRawAST } from "@tact-lang/compiler/dist/grammar/store";
import { AstStore } from "@tact-lang/compiler/dist/grammar/store";
import { enableFeatures } from "@tact-lang/compiler/dist/pipeline/build";
import { precompile } from "@tact-lang/compiler/dist/pipeline/precompile";
import { createNodeFileSystem } from "@tact-lang/compiler/dist/vfs/createNodeFileSystem";

/**
 * Parses the project defined in the Tact configuration file, generating its AST.
 *
 * @param mistiCtx Misti context
 * @param projectRoot Absolute path to the root the project
 * @param config The Tact configuration object: contents of the existing file or a generated object
 * @returns A mapping of project names to their corresponding ASTs.
 */
export function parseTactProject(
  mistiCtx: MistiContext,
  projectConfig: ConfigProject,
  projectRoot: string,
): AstStore | never {
  const project = createNodeFileSystem(projectRoot, false);
  const stdlibPath = mistiCtx.config.tactStdlibPath ?? getStdlibPath();
  const stdlib = createNodeFileSystem(stdlibPath, false);
  mistiCtx.logger.debug(`Parsing project ${projectConfig.name} ...`);
  try {
    let ctx = new CompilerContext();
    ctx = enableFeatures(ctx, mistiCtx.logger, projectConfig);
    ctx = precompile(ctx, project, stdlib, projectConfig.path);
    return getRawAST(ctx);
  } catch (error: unknown) {
    throw TactException.make(error);
  }
}
