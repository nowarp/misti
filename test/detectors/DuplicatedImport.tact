import "@stdlib/ownable";
import "@stdlib/ownable"; // Bad: Duplicated import

import "./NeverAccessedVariables";
import "./NeverAccessedVariables.tact"; // Bad: Duplicated import
