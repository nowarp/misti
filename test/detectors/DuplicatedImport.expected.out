[MEDIUM] DuplicatedImport: Duplicated import
test/detectors/DuplicatedImport.tact:2:1:
  1 | import "@stdlib/ownable";
> 2 | import "@stdlib/ownable"; // Bad: Duplicated import
      ^~~~~~~~~~~~~~~~~~~~~~~~~
  3 | 
`/home/jubnzv/Dev/misti/node_modules/@tact-lang/compiler/dist/stdlib/stdlib/libs/ownable.tact` is already imported in this file
Help: Remove this import statement
See: https://nowarp.io/tools/misti/docs/detectors/DuplicatedImport

[MEDIUM] DuplicatedImport: Duplicated import
test/detectors/DuplicatedImport.tact:5:1:
  4 | import "./NeverAccessedVariables";
> 5 | import "./NeverAccessedVariables.tact"; // Bad: Duplicated import
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  6 | 
`/home/jubnzv/Dev/misti/test/detectors/NeverAccessedVariables.tact` is already imported in this file
Help: Remove this import statement
See: https://nowarp.io/tools/misti/docs/detectors/DuplicatedImport